macro/src/expand.rs

```diff
@@ -1381,6 +1518,122 @@ fn expand_rust_vec(key: NamedImplKey, types: &Types, explicit_impl: Option<&Impl
+            let elem = key.rust;
+            if types.try_resolve(key.rust).is_none() {
+                return TokenStream::new();
+            }
+            let resolve = types.resolve(elem);
+            let link_prefix = format!("cxxbridge1$rust_option$const${}$", resolve.name.to_symbol());
+            let local_prefix = format_ident!("{}__const__ref__option_", elem);
+            (elem, key, resolve, link_prefix, local_prefix)
+        }
+        OptionInner::MutRef(key) => {
+            let elem = key.rust;
+            if types.try_resolve(key.rust).is_none() {
+                return TokenStream::new();
+            }
+            let resolve = types.resolve(elem);
+            let link_prefix = format!("cxxbridge1$rust_option${}$", resolve.name.to_symbol());
+            let local_prefix = format_ident!("{}__ref__option_", elem);
+            (elem, key, resolve, link_prefix, local_prefix)
+        }
+    };
+    let link_new = format!("{}new", link_prefix);
+    let link_drop = format!("{}drop", link_prefix);
+    let link_has_value = format!("{}has_value", link_prefix);
+    let link_value_ptr = format!("{}value_ptr", link_prefix);
+    let link_set = format!("{}set", link_prefix);
+
+    let local_new = format_ident!("{}new", local_prefix);
+    let local_drop = format_ident!("{}drop", local_prefix);
+    let local_has_value = format_ident!("{}has_value", local_prefix);
+    let local_value_ptr = format_ident!("{}value_ptr", local_prefix);
+    let local_set = format_ident!("{}set", local_prefix);
+
+    let begin_span = explicit_impl.map_or(key.begin_span, |explicit| explicit.impl_token.span);
+    let end_span = explicit_impl.map_or(key.end_span, |explicit| explicit.brace_token.span.join());
+    let unsafe_token = format_ident!("unsafe", span = begin_span);
+
+    let ident = key.rust;
+    let (impl_generics, ty_generics) = generics::split_for_impl(*key, explicit_impl, resolve);
+    let (ty, ty_ptr, prevent_unwind_drop_label) = match inner {
+        OptionInner::RustBox(_) => (
+            quote! { ::cxx::alloc::boxed::Box<#elem #ty_generics> },
+            quote! { ::cxx::alloc::boxed::Box<#elem #ty_generics> },
+            format!("::alloc::boxed::Box<::{}> as Drop>::drop", ident),
+        ),
+        OptionInner::Ref(_) => (
+            quote! { &#elem #ty_generics },
+            quote! { *const #elem #ty_generics },
+            format!("&::{}> as Drop>::drop", ident),
+        ),
+        OptionInner::MutRef(_) => (
+            quote! { &mut #elem #ty_generics },
+            quote! { *mut #elem #ty_generics },
+            format!("&mut ::{}> as Drop>::drop", ident),
+        ),
+    };
+    let set_impl = match inner {
+        OptionInner::RustBox(_) => quote! {
+            #[doc(hidden)]
+            #[export_name = #link_set]
+            unsafe extern "C" fn #local_set #impl_generics(this: *mut ::cxx::private::RustOption<#ty>, value: *mut ::core::mem::MaybeUninit<#ty>) {
```

Comment:
Here is the actual implementation side of the spooky rvalue reference I mention above, it causes me to take in something line `*mut MaybeUninit<&T>` which feels weird. Likely I should just be taking in `MaybeUninit<&T>` (by passing the pointer by value)